alerting/src/main/kotlin/org/opensearch/alerting/MonitorRunner.kt

```diff
@@ -757,39 +759,119 @@ object MonitorRunner : JobRunner, CoroutineScope, AbstractLifecycleComponent() {
+                    // TODO: check if need to create alert, if so create it and point it to FindingId
+                    // TODO: run action as well, but this mat need to be throttled based on Mo's comment for bucket level alerting
+                }
+            }
+        }
+
+        // TODO: Check for race condition against the update monitor api
+        // This does the update at the end in case of errors and makes sure all the queries are executed
+        val updatedMonitor = monitor.copy(lastRunContext = updatedLastRunContext)
+        // note: update has to called in serial for shards of a given index.
+        // make sure this is just updated for the specific query or at the end of all the queries
+        updateMonitor(client, xContentRegistry, settings, updatedMonitor)
+    }
+
+    private fun createFindings(
+        monitor: Monitor,
+        index: String,
+        docLevelQuery: DocLevelQuery,
+        matchingDocIds: Set<String>,
+        trigger: DocumentLevelTrigger
+    ): String {
+        val finding = Finding(
+            id = UUID.randomUUID().toString(),
+            relatedDocId = matchingDocIds.joinToString(","),
+            monitorId = monitor.id,
+            monitorName = monitor.name,
+            index = index,
+            queryId = docLevelQuery.id,
+            queryTags = docLevelQuery.tags,
+            severity = docLevelQuery.severity,
+            timestamp = Instant.now(),
+            triggerId = trigger.id,
+            triggerName = trigger.name
+        )
+
+        val findingStr = finding.toXContent(XContentBuilder.builder(XContentType.JSON.xContent()), ToXContent.EMPTY_PARAMS).string()
+        // change this to debug.
+        logger.info("Findings: $findingStr")
+
+        // todo: below is all hardcoded, temp code and added only to test. replace this with proper Findings index lifecycle management.
+        val indexRequest = IndexRequest(".opensearch-alerting-findings")
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .source(findingStr, XContentType.JSON)
+
+        client.index(indexRequest).actionGet()
+        return finding.id
     }
 
-    private suspend fun runForEachQuery(monitor: Monitor, lastRunContext: MutableMap<String, Any>, index: String, query: DocLevelQuery) {
+    private suspend fun runForEachQuery(
+        monitor: Monitor,
+        lastRunContext: MutableMap<String, Any>,
+        updatedLastRunContext: MutableMap<String, Any>,
+        index: String,
+        query: DocLevelQuery
+    ): Set<String> {
         val count: Int = lastRunContext["shards_count"] as Int
+        val matchingDocs = mutableSetOf<String>()
         for (i: Int in 0 until count) {
             val shard = i.toString()
             try {
                 logger.info("Monitor execution for shard: $shard")
 
-                val maxSeqNo: Long = getMaxSeqNo(index, shard)
+                val maxSeqNo: Long = updatedLastRunContext[shard].toString().toLong()
```

Comment:
I added a new object, but note that this object will be modified more later on as we optimize the flow and utilize the object. I didn't see any benefit to add the matchingDocs, so that was not added.